### DISCUSSION TITLE
Could multiple interlocutors be impersonated on both AI and User side?

### DISCUSSION THREAD

**Discussion by ARandomUserFromGithub** [oobabooga/text-generation-webui#485] [Ideas]
I created a character based on an epic bookwriter from the 19th century.
It suddenly introduced 2 more characters, his mother and his father.
It would be difficult to have these characters properly dialogue.
It seems that when I write interlocutor names between {} symbols, as {interlocutor} and {AICharacter}, that the AI understands who it refers to.

I would like that when I write {Interlocutor 2} or {my friend trying AI with my own keyboard} like "{interlocutor 2} says:" and write the quote between quotes like {"Hi, AICharacter, can you tell me more about yourself?"}... I would like the dialogue to use the name tag and create a 2nd user which can be invoked this way and I would like the AI to differentiate between both interlocutors and give different inputs in relation to what it learned about each interlocutor without confusing, but also keeping an understanding that both interlocutors knows, X and don't know Y about their co-conversationalist.

I would like the AI to be capable of impersonating new characters it creates and have these impersonations be displayed with different avatars using the first letter of their name which could be replaced from a character list.

I would like the user to be capable of indicating actions done by the AI and the impersonations on both sides of the AI/User threshold.

If I say *{interlocutor 1} and {interlocutor 2} enters the room.* between asterisks, it would interpret it as 2 human interlocutors entering an imaginary room. 

Characters that are created on the fly would be played by the party who created it, principally, unless the other speaking party would like to impersonate it, to add more punch to the story. 

When {AICharacter 1} creates {AICharacter 2}, it impersonates each one depending on who {interlocutor(s)} addresses to.

Separate AI Characters could submit separate inputs and talk to each other down to logical closure, with a formulation that ends up soon enough by an invitation to {interlocutor(s)} to provide more inputs.

{interlocutor 2} says: "Hi, {AICharacter 2}, how do you relate to {AICharacter 1}?"
{AICharacter 1} interrupts: "We are not related at all, I'm outa here"
{AICharacter 2 retorts: "That'S not true, you're my wife's brother and you still owe me 20$"}
{AICharacter 1} replies: "That's not true, I paid you back last christmas and {Interlocutor 1} and his brother {interlocutor 2} can bare witness. Isn't that true, Interlocutors?
{Interlocutor 2} replies: "That's not even true, it was monopoly money, and you had it by stealing it from the game bank when you ended up broke only at the 4th round"

Newly created characters could have random personalities generated according to context and be created as complex characters which could be invoked in different conversations by importing them from a generated characters folder.

Also, it would be important to have the ability to create user characters where I could get {interlocutor 2} to be an impersonation of Chuck Norris that I would act, textually.

Any user-acted characters could be invoked during a chat with the AI and its own impersonations.

I believe it is necessary in order to create more complex inspirational stories to pop out and inspire some fantastic adventures to participate in interactively.

One last character that would be of importance is some kind of a gamemaster, a narrator that narrates contextual information and calls the various characters when necessary.

**Comment by mcmonkey4eva** [comment]
tldr because that post is long:

The suggestion is, a way to load multiple character files at once, and continue which the bot speaks as and which the user speaks as at any given time.

**Reply by ARandomUserFromGithub** [reply]
It's more intricate than that. 
SE-RA (Short enough, Read Again)
The idea is to have user profiles, and to have the ability to have multiple connected users at once, and multiple impersonated characters at once that can interact together and wait for user inputs before giving new replies.
Finally, a crucial feature is the capability for the AI and the USER to create new characters on the fly using this formulation: `*{{User-Jimmy}} enters the room*` or `*AICharacter-Leon enters the room*` would create a new user or a new AI character and the characters could be described further in a separate editor but could also be prompted to add new details by user and AI, AI being only capable of creating and editing AI characters as it self-prompts it, but user having all permissions to edit both sides.

Another way user could prompt the creation of a new character could be this: *Frank enters the room. Frank is a character impersonated by AI and is good at making riddles and telling the user if his answers are wrong or right and tell the response if the user's response isn't right. Frank says: "Hello, {{user}}* and it would create a new AI character for the AI side of things to impersonate and it would load it inside the discussion and make it available for both user and AI to mention it or even impersonate it.

**Comment by ARandomUserFromGithub** [comment]
Additional info : 
What a User Character Sheet could include is information like: "I am USER, I have a nice sense of humour, and I'm AI CHARACTER 1 and AI Character 2's biggest fan and I came to learn more about their hopes and dreams"
Characters are impersonated only one at once in my scenario
Also AI could ask questions between characters and have them reply to each other and ask USERS to participate
Also as it probably can have multiple connected users at once, it could have multiple user-impersonated characters with their own character sheets
In order, the priority would probably to have a user character sheet first, then multiple ones, then going ahead with more complex functions
If anyone want to participate in this, the thread is open
Character data is just initialisation text, you can hint the bot's personality as you chat with it, and so you can hint your own personality as you chat. Initially, giving user description happens when you prompt your characteristics for the AI to consider them,
Having a user character sheet remembers your character name, and it could contain initialisation text after the <start> of the bot as an {{user}} entry

------

User impersonating a 2nd user could use a special format to indicate which character speaks, and it could be formatted accordingly
Tho when things gets more intricate, migrating 1 side of X and 1 side of Y method could be deployed in multiple-side-of-x and multiple-sides-of-y later, with each characters, AI or USer having their individuated bubbles
So... The memory addon could have multiple slots for multiple characters and while they would have 1 fixed name, the AI could be capable to recognised them using a wider variety of triggers, like John Lemon referred as "mister Lemon" would associate to the right character.
Also, one very important factor I mentioned in my post: It's important that the AI and the User can create new characters on the fly, which would generate new json files
Example: `*{{User-Jimmy}} enters the room*` would create a new user called jimmy if it doesn't already exist. When impersonating Jimmy, the user could say {{user-Jimmy}} replies, with a surprised look on his face:"quote by Jimmy" and I assume that a javascript could help to include the right user tags from a menu when necessary.

`*AICharacter-Leon enters the room*` would create a new AI character and the characters could be described further in a separate editor but could also be prompted to add new details by user and AI, AI being only capable of creating and editing AI characters as it self-prompts it, but user having all permissions to edit both sides.

Another way user could prompt the creation of a new character could be this: `*Frank enters the room. Frank is a character impersonated by AI and is good at making riddles and telling the user if his answers are wrong or right and tell the response if the user's response isn't right. Frank says: "Hello, {{user}}*` and it would create a new AI character for the AI side of things to impersonate and it would load it inside the discussion and make it available for both user and AI to mention it or even impersonate it.
